### ISSUE TITLE
Clone derive macro's trait bound checker issues false positive on generic type parameter

### ISSUE THREAD

**Issue by tmvkrpxl0** [rust-lang/rust#109585]
I tried this code:

```rust
use std::cell::RefCell;
use std::hash::Hash;
use std::rc::Rc;

#[derive(Clone)]
pub struct Container {
    pub object: SharedObject<Dummy>
}

#[derive(PartialEq, Eq, Clone)]
pub struct SharedObject<T:Hash + Eq+ PartialEq>(pub Rc<RefCell<T>>);

#[derive(Hash, Eq, PartialEq)]
pub struct Dummy {}
```

I expected to see this happen: I expected it to compile just fine

Instead, this happened: It fails with compilation error 'the trait bound `Dummy: Clone` is not satisfied'

### Meta

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (8be3c2bda 2023-03-24)
binary: rustc
commit-hash: 8be3c2bda6b683f87b24714ba595e8b04faef54c
commit-date: 2023-03-24
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

<details><summary>Backtrace</summary>
<p>

```
RUST_BACKTRACE=1 cargo build
   Compiling lisp_parser v0.1.0 (/home/tmvkrpxl0/CLionProjects/ExperimentalLanguage)
warning: unused import: `BitOr`
 --> src/types.rs:3:24
  |
3 | use std::ops::{BitAnd, BitOr};
  |                        ^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0277]: the trait bound `Node: Clone` is not satisfied
 --> src/types.rs:8:10
  |
6 | #[derive(Eq, PartialEq, Clone)]
  |                         ----- in this derive macro expansion
7 | pub enum Type {
8 |     Unit(SharedObject<Node>),
  |          ^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Node`
  |
note: required for `SharedObject<Node>` to implement `Clone`
 --> src/ast.rs:9:25
  |
9 | #[derive(PartialEq, Eq, Clone)]
  |                         ^^^^^ unsatisfied trait bound introduced in this `derive` macro
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `Pointer`
 --> src/syntax_analyzer.rs:3:36
  |
3 | use std::fmt::{Display, Formatter, Pointer};
  |                                    ^^^^^^^

For more information about this error, try `rustc --explain E0277`.
warning: `lisp_parser` (lib) generated 2 warnings
error: could not compile `lisp_parser` (lib) due to previous error; 2 warnings emitted
```

</p>
</details>

The purpose of `SharedObject` in my project is wrapper for Rc<RefCell<T>>. type alias may also work, but I've implemented Hash trait for using it as a key for HashMap. T in this context represent type that are big and should avoid being copied.
I expect it to work well considering Rc is clone-able type.
Implementing Clone manually instead works:
```rs
use std::cell::RefCell;
use std::hash::Hash;
use std::rc::Rc;

#[derive(Clone)]
pub struct Container {
    pub object: SharedObject<Dummy>
}

#[derive(PartialEq, Eq)]
pub struct SharedObject<T:Hash + Eq+ PartialEq>(pub Rc<RefCell<T>>);

impl<T: Hash + Eq+ PartialEq> Clone for SharedObject<T> {
    fn clone(&self) -> Self {
        SharedObject(self.0.clone())
    }
}

#[derive(Hash, Eq, PartialEq)]
pub struct Dummy {}
```

After looking through derive macro of Clone, I found that it expects all type parameters to also implement Clone. I find it bit strange, as wrapping it with Rc would work just fine.

**Comment by jyn514** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/26925.
